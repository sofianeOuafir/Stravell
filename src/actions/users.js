import { editAuthUserPhotoURL } from "./auth";
import { editPostsUserPhotoURL } from "./posts";
import database from "./../firebase/firebase";

export const startUpdateUserProfilePicture = ({
  uid,
  userPhotoURL
}) => async dispatch => {
  let updateObject = {};

  // users/uid
  updateObject[`users/${uid}/userPhotoURL`] = userPhotoURL;

  // user-posts
  await database.ref(`user-posts/${uid}/`).once("value", function(snapshot) {
    let postKeys;
    try {
      postKeys = Object.keys(snapshot.val());
    } catch (error) {
      postKeys = [];
    }

    postKeys.forEach(key => {
      updateObject[`user-posts/${uid}/${key}/userPhotoURL`] = userPhotoURL;
    });
  });

  // posts
  await database
    .ref(`posts/`)
    .orderByChild("uid")
    .equalTo(uid)
    .once("value", function(snapshot) {
      let postKeys;
      try {
        postKeys = Object.keys(snapshot.val());
      } catch (error) {
        postKeys = [];
      }
      postKeys.forEach(key => {
        updateObject[`posts/${key}/userPhotoURL`] = userPhotoURL;
      });
    });

  // country-posts
  await database.ref(`country-posts/`).once("value", function(snapshot) {
    snapshot.forEach(function(child) {
      const countryCode = child.key;
      child.ref
        .orderByChild("uid")
        .equalTo(uid)
        .once("value", function(snapshot) {
          snapshot.forEach(function(child) {
            const postId = child.key;
            updateObject[
              `country-posts/${countryCode}/${postId}/userPhotoURL`
            ] = userPhotoURL;
          });
        });
    });
  });

  // update place-posts
  await database.ref(`place-posts/`).once("value", function(snapshot) {
    snapshot.forEach(function(child) {
      const placeId = `${child.key}`;
      child.ref
        .orderByChild("uid")
        .equalTo(uid)
        .once("value", function(snapshot) {
          snapshot.forEach(function(child) {
            const postId = child.key;
            updateObject[
              `place-posts/${placeId}/${postId}/userPhotoURL`
            ] = userPhotoURL;
          });
        });
    });
  });

  // // update region-posts
  await database.ref(`region-posts/`).once("value", function(snapshot) {
    snapshot.forEach(function(child) {
      const regionCode = child.key;
      child.ref
        .orderByChild("uid")
        .equalTo(uid)
        .once("value", function(snapshot) {
          snapshot.forEach(function(child) {
            const postId = child.key;
            updateObject[
              `region-posts/${regionCode}/${postId}/userPhotoURL`
            ] = userPhotoURL;
          });
        });
    });
  });

  return database
    .ref()
    .update(updateObject)
    .then(() => {
      dispatch(editAuthUserPhotoURL(`${userPhotoURL}?${new Date().getTime()}`));
      dispatch(
        editPostsUserPhotoURL({
          userPhotoURL: `${userPhotoURL}?${new Date().getTime()}`,
          uid
        })
      );
    })
    .catch(e => {
      console.log(e);
    });
};
