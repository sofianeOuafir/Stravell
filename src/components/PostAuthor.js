import Link from "next/link";
import React from "react";
import Avatar from "react-avatar";

const PostAuthor = ({ authorUid, avatarSize, authorPhotoURL, authorName }) => (
  <Link
    href={`/user?uid=${authorUid}`}
  >
    <div>
      <Avatar
        className="mr1"
        size={avatarSize}
        round={true}
        src={authorPhotoURL}
      />
      <span className="c-dark-grey">{authorName}</span>
    </div>
  </Link>
);

export default PostAuthor;
