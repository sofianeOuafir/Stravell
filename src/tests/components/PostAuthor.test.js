import React from 'react';
import { shallow } from 'enzyme';
import PostAuthor from './../../components/PostAuthor';
import uuid from 'uuid';
import users from './../fixtures/users';

test('should render correctly', () => {
  const { uid: authorUid, userPhotoURL: authorPhotoURL, userName: authorName, avatarSize = 40 } = users[0];
  const props = { authorUid, authorPhotoURL, authorName, avatarSize };
  const wrapper = shallow(<PostAuthor { ...props } />);
  expect(wrapper).toMatchSnapshot();
});