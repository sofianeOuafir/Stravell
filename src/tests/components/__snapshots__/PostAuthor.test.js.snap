// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Link
  as="/u/show/samia-ouafir/1pUNibuDtBPmdsBPpAqlNEZYc9r2"
  href="/user?uid=1pUNibuDtBPmdsBPpAqlNEZYc9r2"
>
  <a
    className="no-text-decoration"
  >
    <div>
      <withAvatarConfig
        className="mr1"
        round={true}
        size={40}
        src="https://graph.facebook.com/2307269322671171/picture"
      />
      <span
        className="c-dark-grey"
      >
        Samia Ouafir
      </span>
    </div>
  </a>
</Link>
`;
